Replace HttpStreamsHandler for netty HTTP client

yawkat opened the pull request:
Just like #9128 did for the server, this PR replaces the HttpStreamsHandler and HandlerPublisher for the client. Instead, incoming HttpRequest and HttpContent objects are handled directly and transformed to the needed types.

This allows full removal of the streams handler. Some of the functionality is reimplemented in ReactiveClientReader and ReactiveClientWriter, but these are much simpler classes and only need to be used for streaming requests / responses.

This implementation is not quite as optimized as PipeliningServerHandler, there's no read coalescing for example, and it uses multiple handlers where a more complex implementation could do with one.

yawkat committed:
Replace HttpStreamsHandler for netty HTTP client
Just like #9128 did for the server, this PR replaces the HttpStreamsHandler and HandlerPublisher for the client. Instead, incoming HttpRequest and HttpContent objects are handled directly and transformed to the needed types.

This allows full removal of the streams handler. Some of the functionality is reimplemented in ReactiveClientReader and ReactiveClientWriter, but these are much simpler classes and only need to be used for streaming requests / responses.

This implementation is not quite as optimized as PipeliningServerHandler, there's no read coalescing for example, and it uses multiple handlers where a more complex implementation could do with one.

graemerocher commented:
it would be good to re-enable binary compatible checks and rebase this PR so we get a clearer idea of what the binary incompatible changes are

yawkat commented:
with #9478 there should be none, it's only internal classes now.

graemerocher commented:
@yawkat please rebase

yawkat committed:
Merge branch '4.1.x' into client-streams

# Conflicts:
#	http-netty/src/main/java/io/micronaut/http/netty/LastHttp2Content.java
#	http-netty/src/main/java/io/micronaut/http/netty/stream/DefaultHttp2Content.java
#	http-netty/src/main/java/io/micronaut/http/netty/stream/DefaultLastHttp2Content.java
#	http-netty/src/main/java/io/micronaut/http/netty/stream/Http2Content.java
#	http-netty/src/main/java/io/micronaut/http/netty/stream/HttpStreamsHandler.java

yawkat committed:
Consistent errors for closed channels (#9652)
Channel close during the initial response headers or during the response body now raises a new, consistent ResponseClosedException.

yawkat commented:
I've added a change to fix the brittle test reported in #9652.

msupic committed:
Merge branch '4.1.x' of github.com:micronaut-projects/micronaut-core into client-streams

sdelamo commented on http-client-tck/src/main/java/io/micronaut/http/client/tck/tests/ContinueTest.java:
```diff
@@ -0,0 +1,125 @@
+        );
+    }
+
+    @Controller("/continue")
```
Please, annotate the controller with `@Requires(property = spec.name` to avoid bean pollution between tests

sdelamo commented on http-netty/src/test/groovy/io/micronaut/http/netty/reactive/HandlerPublisherSpec.groovy:
```diff
@@ -1,88 +0,0 @@
-import org.reactivestreams.Subscription
-import spock.lang.Specification
-
-class HandlerPublisherSpec extends Specification {
```
do we test this somewhere else?

yawkat replied:
HandlerPublisher is gone now, the reactive publisher doesn't buffer so it's not susceptible

sdelamo commented on http-netty/src/test/groovy/io/micronaut/http/netty/stream/HttpStreamsClientHandlerSpec.groovy:
```diff
@@ -1,66 +0,0 @@
-
-class HttpStreamsClientHandlerSpec extends Specification {
-    @Issue('https://github.com/micronaut-projects/micronaut-tracing/issues/316')
-    def 'out of order write'() {
```
do we test this somewhere else?

yawkat replied:
content isnt delayed anymore so the test case is obsolete

yawkat committed:
add requires

yawkat committed:
Merge remote-tracking branch 'origin/client-streams' into client-streams
